woocommerce/woocommerce-gateway-stripe
# Process 3DS modal success requests when changing a subscriptions payment method

Fixes #3312

## Changes proposed in this Pull Request:

In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3307, I changed setup intent requests to make sure we use the Stripe SDK to complete the next actions (eg 3D-Secure challenges). This means the modal pops up rather than a redirect. 

This inadvertently broke the change payment method flow for subscriptions because after completing the 3D-Secure challenge, the JS attempts to confirm the order payment. Because the "order" in this context is actually a subscription, you would get 2 issues:

1. The intent validation would fail. 
2. The subscription would be set to "pending". 

This PR fixes that by carving out a flow for change payment method requests.

> [!note]
> https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3307 is unreleased so this error doesn't currently impact the public version of the plugin.

## Testing instructions

1. Enable WooCommerce Subscriptions
2. Purchase a subscription product.
3. From the **My Account > Subscriptions** page view the subscription.
4. Click the Change payment button.
5. On the change payment page enter a 3D-secure enabled card. eg `4000002760003184`.
6. The page will do a brief redirect before the 3D-Secure challenge modal will appear.
7. Complete the auth.
   - On `develop` you will see the following error and the subscription will be "pending". 

| Error message | Subscription "Pending" and Note |
|--------|--------|
| <img width="700" alt="Screenshot 2024-07-25 at 2 11 18 PM" src="https://github.com/user-attachments/assets/c3945280-ea94-42e4-91a2-b311af3c6628"> | <img width="700" alt="Screenshot 2024-07-25 at 2 11 38 PM" src="https://github.com/user-attachments/assets/6747472f-e4fc-4b98-849d-e1843a74e2ad"><img width="200" alt="Screenshot 2024-07-25 at 2 12 41 PM" src="https://github.com/user-attachments/assets/823e9e45-379b-4458-a0c7-5d0333bd44cc"> | 

8. Repeat the steps on this branch and the subscription change payment request should complete as expected. 



---


**Post merge**
